courses/models.py

```diff
@@ -0,0 +1,119 @@
+from django.db import models
+from utils.constants import SKILL_BEGINNER, SKILL_INTERMEDIATE, SKILL_ADVANCED, EXTERNAL_LINK_BLOG, EXTERNAL_LINK_TUTORIAL, EXTERNAL_LINK_COURSE
+from django.utils.text import slugify
+# Create your models here.
+
+class Language(models.Model):
+    language_name = models.CharField(max_length=20)
+    slug = models.SlugField(unique=True)
+
+    def _get_unique_slug(self):
+        slug = slugify(self.language_name)
+        unique_slug = slug
+        num = 1
+        while Language.objects.filter(slug=unique_slug).exists():
+            unique_slug = '{}-{}'.format(slug, num)
+            num += 1
+        return unique_slug
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = self._get_unique_slug()
+        super(Language, self).save(*args, **kwargs)
+
+    def __str__(self):
+        return '{name}-{slug}'.format(name=self.language_name, slug=self.slug)
+
+
+class Domain(models.Model):
+    domain_name = models.CharField(max_length=20)
+    slug = models.SlugField(unique=True)
+
+    def _get_unique_slug(self):
+        slug = slugify(self.domain_name)
+        unique_slug = slug
+        num = 1
+        while Domain.objects.filter(slug=unique_slug).exists():
+            unique_slug = '{}-{}'.format(slug, num)
+            num += 1
+        return unique_slug
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = self._get_unique_slug()
+        super(Domain, self).save(*args, **kwargs)
+
```

Comment:
similar code block for rest of the case. Think about creating a decorator/wrapper. This will also give you hands-on experience writing custom decorators and teach you another layer of code abstraction